src/python/pants/init/specs_calculator.py

```diff
@@ -69,27 +76,57 @@ def create(
         exclude_patterns: Optional[Iterable[str]] = None,
         tags: Optional[Iterable[str]] = None,
     ) -> Specs:
+        # Determine the literal specs.
         specs = cls.parse_specs(
             raw_specs=options.specs,
             build_root=build_root,
             exclude_patterns=exclude_patterns,
             tags=tags,
         )
 
+        # Determine `Changed` arguments directly from options to support pre-`Subsystem`
+        # initialization paths.
         changed_options = ChangedOptions.from_options(options.for_scope("changed"))
 
+        # Parse --query expressions into objects which can be resolved into BuildFileAddresses via v2
+        # rules.
+        query_expr_strings = options.for_global_scope().query
+        exprs = session.product_request(
```

Comment:
This call being a synchronous call to the engine feels extraneous... it feels like all of query parsing could be pushed down under the second call to `product_request` (for `QueryAddresses`) in this PR, and rather than checking whether `exprs` is non-null, you could just check whether the query is empty or not.

Reply:
I've added a TODO regarding this -- I wasn't able to remove the first synchronous call yet because of the lack of the ability to return a `@union` type right now, so the `QueryComponentWrapper` is a workaround for type erasure right now. This should be improved in #7356!